src/polygonParts/models/polygonPartsManager.ts

```diff
@@ -198,48 +212,172 @@ export class PolygonPartsManager {
+        when array_length(${filterRequestFeatureIds}, 1) = 1 then jsonb_build_object('${requestFeatureId}', ${filterRequestFeatureIds}[1])
+        else jsonb_build_object('${requestFeatureId}', ${filterRequestFeatureIds})
+      end
+    )`
+      : '';
+
+    const findPolygonPartsSelect = entityManager.createQueryBuilder().select(
+      `jsonb_build_object(
           'type', 'FeatureCollection',
           'features', coalesce(jsonb_agg(
             jsonb_build_object(
               'type', 'Feature',
               'geometry', st_asgeojson(${geometryColumn}, 15)::jsonb,
-              'properties', to_jsonb(clip) - '${geometryColumn}'
+              'properties', to_jsonb(${filterQueryAlias}) - '{${geometryColumn},${filterRequestFeatureIds}}'::text[] ${requestFeatureIds}
             )
           ), '[]')
         ) AS ${'geojson' satisfies keyof FindPolygonPartsQueryResponse}`
+    );
+    return findPolygonPartsSelect;
+  }
+
+  private buildPolygonPartsQueryWithFilter<T extends ObjectLiteral, ShouldClip extends boolean = boolean>(
+    context: FindPolygonPartsOptions<ShouldClip> & {
+      entityManager: EntityManager;
+      select: SelectQueryBuilder<T>;
+      filterQuery: { filterQueryAlias: string; filterRequestFeatureIds: string; findSelectOutputColumns: string[] };
+    }
+  ): SelectQueryBuilder<T> {
+    const {
+      entityManager,
+      filter,
+      filterQuery: { filterQueryAlias, filterRequestFeatureIds, findSelectOutputColumns },
+      polygonPartsEntityName,
+      select,
+      shouldClip,
+    } = context;
+    const hasAnyGeometries = filter.features.some((feature) => feature.geometry);
+    const canClip = hasAnyGeometries && shouldClip;
+
+    const polygonPart = entityManager.getRepository(PolygonPart);
+    polygonPart.metadata.tablePath = polygonPartsEntityName.databaseObjectQualifiedName; // this approach may be unstable for other versions of typeorm - https://github.com/typeorm/typeorm/issues/4245#issuecomment-2134156283
+
+    const filterExtractGeometriesCTE = entityManager
+      .createQueryBuilder()
+      .select(`jsonb_array_elements('${JSON.stringify(filter)}'::jsonb -> 'features')`, 'filter_feature')
+      .addSelect(`jsonb_array_elements('${JSON.stringify(filter)}'::jsonb -> 'features') ->> 'geometry'`, 'filter_geometry')
+      .fromDummy();
+
+    const filterNonNullGeometriesCTE = entityManager
+      .createQueryBuilder()
+      .select(['filter_feature', 'st_geomfromgeojson(filter_geometry) as filter_geometry'])
+      .from('filter_extract_geometries', 'filter_extract_geometries')
+      .where('filter_geometry is not null');
+
+    const filterNonNullGeometriesCounterCTE = entityManager
+      .createQueryBuilder()
+      .select('count(1)', 'count')
+      .from('filter_non_null_geometries', 'filter_non_null_geometries');
+
+    const filterNullGeometriesCTE = `select filter_extract_geometries.filter_feature, st_makeenvelope(-180, -90, 180, 90, 4326) as filter_geometry
```

Comment:
why do we need this CTE filter if there are null geometries? you already filtered the NonNullGeometries

Reply:
it's a fallback to a polygon of the entire globe in case feature collection is empty or all geometries are null